server/src/main/java/org/elasticsearch/reservedstate/service/AbstractFileWatchingService.java

```diff
@@ -0,0 +1,394 @@
+ *
+ * <p>What this class does not do is define what should happen after the file changes.
+ * An implementation class should override {@link #processFileChanges(Path)} to define
+ * the correct behavior.</p>
+ */
+public abstract class AbstractFileWatchingService extends AbstractLifecycleComponent implements ClusterStateListener {
+
+    private static final Logger logger = LogManager.getLogger(FileSettingsService.class);
+    protected static final int REGISTER_RETRY_COUNT = 5;
+    protected final ClusterService clusterService;
+    protected final ReservedClusterStateService stateService;
+    protected final Path watchedFileDir;
+    protected final Path watchedFile;
+    protected final List<FileChangedListener> eventListeners;
+    protected WatchService watchService; // null;
+    protected Thread watcherThread;
+    protected FileSettingsService.FileUpdateState fileUpdateState;
+    protected WatchKey settingsDirWatchKey;
+    protected WatchKey configDirWatchKey;
+    protected volatile boolean active = false;
+
+    public AbstractFileWatchingService(
+        ClusterService clusterService,
+        ReservedClusterStateService stateService,
+        Environment environment,
+        String watchedDirName,
+        String watchedFileName
+    ) {
+        this.clusterService = clusterService;
+        this.stateService = stateService;
+        this.watchedFileDir = environment.configFile().toAbsolutePath().resolve(watchedDirName);
+        this.watchedFile = watchedFileDir.resolve(watchedFileName);
+        this.eventListeners = new CopyOnWriteArrayList<>();
+    }
+
+    /**
+     * Any implementation of this class must implement this method in order
+     * to define what happens once the watched file changes.
+     */
+    abstract PlainActionFuture<Void> processFileChanges(Path path);
+
+    /**
+     * There may be an indication in cluster state that the file we are watching
+     * should be re-processed: for example, after cluster state has been restored
+     * from a snapshot. By default, we do nothing, but this method should be overridden
+     * if different behavior is desired.
+     * @param clusterState State of the cluster
+     * @return false, by default
+     */
+    protected boolean shouldRefreshFileState(ClusterState clusterState) {
+        return false;
+    }
+
+    /**
+     * 'Touches' the settings file so the file watcher will re-processes it.
+     * <p>
+     * The file processing is asynchronous, the cluster state or the file must be already updated such that
+     * the version information in the file is newer than what's already saved as processed in the
+     * cluster state.
+     *
+     * For snapshot restores we first must restore the snapshot and then force a refresh, since the cluster state
+     * metadata version must be reset to 0 and saved in the cluster state.
+     */
+    protected void refreshExistingFileStateIfNeeded(ClusterState clusterState) {
```

Comment:
Is a subclass expected to be able to override this method, or only the shouldRefreshFileState above (ie could this method be final?)